### ISSUE TITLE
Adding tags to 1% of tests in large test suite causes the test run to take 28% longer

### ISSUE THREAD

**Issue by smaifullerton-wk** [dart-lang/test#1042]
We are unable to use tags in our large repos because of the impact on performance. 

The below times are in a package with almost 3900 tests, compiled with DDC in Dart 2.3.2. I ran the tests with no flags to include or exclude tests tagged tests; the tags are only defined and added to the tests. 

No tags: **3:18**
Tag on one test, no tags defined in dart_test.yaml: **3:40**
Tag on one test, one tag defined in dart_test.yaml: **3:40**
Tag on 20 tests, one tag defined in dart_test.yaml: **4:05**
Tag on 40 tests, two tags defined in dart_test.yaml: **4:13**

The issue seems to be exacerbated when compiled with dart2js - a test run with no tags took 6 minutes, and defining one tag and adding it to 20 tests brought it up to 8.5 minutes.

This issue has existed for at least 7 months, and in Dart 1 as well.

As there's no supported way to determine the compiler at runtime, we were hoping to use tags as a way to skip tests with known dart2js bugs when running tests in release mode, but cannot add this much time to each test run.

I'm happy to try things if you need any other information. Thanks!

**Comment by grouma** [comment]
Thanks for the issue. We'll need a little more information on your test set up. Can you provide an example set of tests? Do your tests have the `tag` attribute or a top level annotation? How are you running your tests? Do you include `-x` to filter tests?

**Comment by smaifullerton-wk** [comment]
@grouma I just took a stab at creating a package that had 4000 simple tests to repro the issue; unfortunately, I didn't get the same results. I'll give it a shot in a big open source repo and see if I can get you a branch.

The runs in the PR description use the `tags` optional parameter with the same tag I specified in dart_test.yaml (via `tags:`). There was no difference between adding it to groups, or an equivalent number of individual tests. 

I just tried using the `@Tags` annotation in a file with ~50 tests, and the test run took 3:36. On several files with around 200 tests, 3:35. 

DDC: `pub run build_runner test -- -p chrome test/unit/tests.dart`
dart2js: `pub run build_runner test -r --define 'build_web_compilers|entrypoint=dart2js_args=["--no-minify"]' -- -p chrome test/unit/tests.dart`

I highly doubt it's relevant, but in the build.yaml we also define a BWC entrypoint to restrict the build to just the test runner file.

I do not include `-x` or any other flag to filter tests; I just run the whole suite with tags defined and added to tests.

I'll try to get back to you soon with a way to repro.
